helper.php

```diff
@@ -0,0 +1,336 @@
+			'parameters' => array(
+				'folder' => "string path of to look up (e.g. ':mynamespace:mypage'), if null use current page.",
+			),
+			'return' => "array the config settings key=>value array, e.g. array('pagesize'=>'A4', 'orientation'=>'protrait' ...)",
+		);
+		$result[] = array(
+			'name'   => 'loadExportConfig',
+			'desc'   => "Returns an array of config settings that are going to be used. This includes checking $INPUT for overrides and using current page's hierarchical settings if enabled.",
+			'return' => "array the config settings key=>value array, e.g. array('pagesize'=>'A4', 'orientation'=>'protrait' ...)",
+		);
+		$result[] = array(
+			'name'   => 'addExportButton',
+			'desc'   => "Adds the export by (hierarchical) settings.",
+			'return' => "void",
+		);
+		$result[] = array(
+			'name'   => 'replyAjax',
+			'desc'   => "Handles calls.",
+			'return' => "void (it handled the call so nothing more to do)",
+		);
+		return $result;
+	}
+	
+	function checkSettingstree(){
+		return !!$this->get_settingstree();
+	}
+
+	function getSettingstreeAdminHtml(){
+		if (!$this->get_settingstree()){ return null;}
+		return $this->settings_helper->showAdmin('dw2pdf',$this->_getFolder(null));
+	}
+	private function _getFolder($folder = null){
+		if (is_null($folder)){
+			global $ID;
+			$folder = $ID;
+		}
+		return ':'.ltrim(strtr($folder,'/',':'),':');
+	}
+
+	function getConf($key,$notset = false, $folder = null){
+		$folder = $this->_getFolder($folder);
+		if (!is_array(@$this->_folder_confs[$folder])){
+			$this->getAllConf($folder);
+		}
+		
+		return (array_key_exists($key, (array)$this->_folder_confs[$folder]) ? $this->_folder_confs[$folder][$key] : $notset);
+	}
+	
+	function getAllConf($folder = null){
+		$folder = $this->_getFolder($folder);
+		if (!is_array(@$this->_folder_confs[$folder])){
+			global $conf;
+			$cnf = $conf['plugin']['dw2pdf'];
+			if ($this->get_settingstree()){
+				$cnf = array_merge($cnf,$this->settings_helper->getConf('dw2pdf',$folder));
+			}
+			if (!is_array($cnf)) {$cnf = array();}	// should never happen if the plugin have configs, and this plugin have configs.
+			$this->_folder_confs[$folder] = $cnf;
+		}
+		return $this->_folder_confs[$folder];
+
+	}
+	
+	// CODE MOVED HERE FROM action.php
+	function addExportButton(&$event){
+        if($this->getConf('showexportbutton') && $event->data['view'] == 'main') {
+			// insert button at position before last (up to top)
+			$event->data['items'] = array_slice($event->data['items'], 0, -1, true) +
+				array('export_pdf' => 
+						  '<li>'
+						  .'<a ' . $this->_exportButtonHtml()	// this is the htmlcode for the opening a tag's href and onclick attributes.
+						  . '"  class="action export_pdf" rel="nofollow" title="' . $this->getLang('export_pdf_button') . '">'
+						  . '<span>' . $this->getLang('export_pdf_button') . '</span>'
+						  . '</a>'
+						  . '</li>'
+				) +
+				array_slice($event->data['items'], -1, 1, true);
```

Comment:
Well, try it ;) 
Although I did try to create clean an logical code with detailed comments on more complex/remote restricted part, but you don't trash a weeks work just before deadline because you suddenly noticed you /may/ misunderstood some concept. (I made these PRs on the clock, now I roughly lived out my time allocated to that, and on private time... I'm either working on something else or well... not in state to work.)

---
For export button: currently there are only 2 options, either use export-config or use the default behavior.
Later the extended template's variables could be set up as well (e.g. header, footer, coverpage, page summary settings that the template allows).
Which make you 2^3 - (2^2 + 1) choices, and should have their own function and place.

But this is also a very good chance to add extension/embedation point, just a few possible scenarios suddenly come in mind:
- bookcreator, which currently needs to create a content page/namespace (!) to manage settings and you can use the export button there.
- you don't want export settings only extended template settings
- you want namespace export / book title popup but otherwise default behavior
- you want to use a plugin that sets encryption / digital signs on exported pdfs
- you want to use a plugin that allows admins/groups/cronjobs to create pdf and plain users only to download that (or submit request for a pdf)
- you want different export buttons depending on context (e.g. you want one-click or croned export to get pdf of a CR-status or bug-fix stories, but you need book export for the user manual)

Since only helpers are designed to cross-plugin usage (`getMethods`?), all extension or embedation points should be in helper. This is to keep future options open: extending lifecycle and allowing (more) people to (more) easily contribute (offering customization variability).

---
For Ajax handling: helpers are also meant to be the cross-type shared-code for the same plugin (hence the name), so common ajax handling should go there as well...
You could create more helpers for one plugin, but then your included helper will need to include an other helper (e.g. one helper for conf, one for ajax, several for export/cross-plugin usage per functionality). Since ajax-based request can't use preregistered closures this may makes a lot of fuss. (Note: on ajax, the entry point is the action-type class of the included helper, and not the includer plugin... see usage choices/overview on https://www.dokuwiki.org/plugin:explorertree )

At least that's the concept I understand from digging in devel pages and forums (and maybe mailgroup, but that is not effectively searchable)... You're a core member of dokuwiki, please correct me if I'm on wrong path, as the good code requires correct (and good) concepts.
(Maybe put this on mailgroup so the others can share their ideas as well?)

---
Anyways: next week I should be able to allocate some hours to review this more deeply and fix/change/reorder it according to the wishes/future plans/specicifications. I would be very helpful if by that time we clean up my possible understanding of concepts.